src/plan/barriers.rs

```diff
@@ -112,9 +133,17 @@ impl<E: ProcessEdgesWork> Barrier for ObjectRememberingBarrier<E> {
     }
 
     #[inline(always)]
-    fn post_write_barrier(&mut self, target: WriteTarget) {
+    fn post_write_barrier(&mut self, target: BarrierWriteTarget) {
+        match target {
+            BarrierWriteTarget::Object(obj) => self.barrier(obj),
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    fn post_write_barrier_slow(&mut self, target: BarrierWriteTarget) {
         match target {
-            WriteTarget::Object(obj) => {
+            BarrierWriteTarget::Object(obj) => {
                 self.enqueue_node(obj);
```

Comment:
Should we just call `self.barrier_slow(obj)` to reduce code duplication?  The only difference is inlining.
```suggestion
                self.barrier_slow(obj);
```

Reply:
Right, the difference is inlining. I was trying to achieve these: 1. make it clear in the code that for the write barrier implementation, which is fastpath (inlined), and which is slowpath (never inlined), and 2. make sure that we do not unnecessarily make another call in Rust when a binding calls the slowpath (from JIT'd code into Rust). Having one `inline(always)` method, and another `inline(never)` method that calls the inlined method can achieve this.

Reply:
Yes.  That makes sense if the intention is better documentation.

To reduce a level of function call (if that's important), however, we may label all those functions as `#[inline(always)]`, including `memory_manager::post_write_barrier` (we may or may not want to add `memory_manager::post_write_barrier`. See the other comment near  `memory_manager::post_write_barrier`). The only functions that are not inlined are the `extern "C"` functions defined in the bindings.

Reply:
There are some differences here: 1. `post_write_barrier` is the full barrier implementation in Rust. For an efficient binding, they probably won't call it anyway. But `post_write_barrier_slow` is the slowpath, a binding that aims performance will call it. 2. With no inline directive, a compiler may choose to inline the function. I was trying to avoid `inline(never)`, for which case a compiler is explicitly told not to inline.

But anyway, I can add `#[inline(always)]` as you suggested.

Reply:
I added `#[inline(always)]` to `post_write_barrier()` in `memory_manager`. To be consistent, I also added `#[inline(always)]` to `alloc` and `post_alloc` in `memory_manager`.

Reply:
What about the code duplication here? Should we call `self.barrier_slow(obj);` or keep it like this?

Reply:
`barrier_slow()` is forced out of line. If we call `barrier_slow()`, there will be always an extra call when the binding calls the slowpath.

Reply:
One solution is to let both `barrier_slow` and `post_write_barrier_slow` call a common function, such as `barrier_slow_impl`, which has `#[inline(always)]`.  But if the post write barrier itself is going to be replaced by a subsuming barrier, I am okay with leaving it as is.

Reply:
> One solution is to let both `barrier_slow` and `post_write_barrier_slow` call a common function, such as `barrier_slow_impl`, which has `#[inline(always)]`.

`enqueue_node()` is the common function with `#[inline(always)]`.

Reply:
OK.  That makes sense, at least for the object remembering barrier.